### PR TITLE
Fix excess property check in union and intersection types

### LINKED ISSUE #5237: Weird behaviour with union types and 'strict' object literals
I'm not sure if this a bug or by design, but it's confusing me:

```typescript
interface Constraint {
  message?: string;
}

var badConstraint: Constraint = {
  messge: "rte" // errors as expected
};

var goodConstraint: Constraint | boolean = {
  messge: "rte" // I expect this to error!
};
```

Can anybody explain what is going on here?

http://www.typescriptlang.org/Playground/#src=interface%20Constraint%20%7B%0D%0A%20%20message%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Avar%20badConstraint%3A%20Constraint%20%3D%20%7B%0D%0A%20%20messge%3A%20%22rte%22%20%2F%2F%20errors%20as%20expected%0D%0A%7D%3B%0D%0A%0D%0Avar%20goodConstraint%3A%20Constraint%20%7C%20boolean%20%3D%20%7B%0D%0A%20%20messge%3A%20%22rte%22%20%2F%2F%20I%20expect%20this%20to%20error!%0D%0A%7D%3B%0D%0A

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #5237.

### CODE REVIEW

File: tests/cases/compiler/objectLiteralExcessProperties.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+interface Cover {
+    color?: string;
+}
+
+var b1: Book = { forword: "oops" };
+
+var b2: Book | string = { foreward: "nope" };
+
+var b3: Book | (Book[]) = [{ foreword: "hello" }, { forwards: "back" }];
```
Comment (DanielRosenwasser) [inline tests/cases/compiler/objectLiteralExcessProperties.ts]: Nice test

**Comment by DanielRosenwasser** [comment]
LGTM

**Comment by mhegazy** [comment]
:+1:

**Comment by ahejlsberg** [comment]
The PR modifies the [rules for excess properties](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3115-excess-properties) as follows:

A source type *S* is considered to have excess properties with respect to a target type *T* if

* *S* is a fresh object literal type,
* *T* is an object-like type, and
* *S* has one or more properties that aren't expected in *T*.

A type *T* is said to be object-like if

* *T* is an object type, or
* *T* a union or intersection type containing at least one-object like type.

A property *P* is said to be expected in a type *T* if one of the following is true:

* *T* is an object type and
  * *T* has a property with the same name as *P*,
  * *T* has a string or numeric index signature,
  * *T* has no properties, or
  * *T* is the global type 'Object'.
* *T* is a union or intersection type and *P* is expected in at least one object-like constituent type of *T*.

File: tests/cases/compiler/objectLiteralExcessProperties.ts (hunk 1 above)
Comment (RyanCavanaugh) [inline tests/cases/compiler/objectLiteralExcessProperties.ts]: You should have phrased that using four words

**Pull request merged**